src/Type/Constant/ConstantArrayType.php

```diff
@@ -602,17 +602,20 @@ public function findTypeAndMethodNames(): array
 
 	public function hasOffsetValueType(Type $offsetType): TrinaryLogic
 	{
-		$offsetType = $offsetType->toArrayKey();
-		if ($offsetType instanceof UnionType) {
+		$offsetArrayKeyType = $offsetType->toArrayKey();
+
+		if ($offsetArrayKeyType instanceof UnionType) {
 			$results = [];
-			foreach ($offsetType->getTypes() as $innerType) {
-				$results[] = $this->hasOffsetValueType($innerType);
+			foreach ($offsetArrayKeyType->getTypes() as $innerType) {
+				if (!$innerType->isSuperTypeOf($offsetType)->yes()) {
```

Comment:
Since numeric-string->toArrayKey returns int|numeric-string, we have a infinite loop without this condition.

Not sure if this is the best fix thought...